### ISSUE TITLE
Better support for calculating gradients/Jacobians within computation graph

### ISSUE THREAD

**Issue by neubig** [clab/dynet#494]
Some training algorithms for neural nets require calculating gradients or Jacobians in the process of calculating a loss function. For example, see line 7 of algorithm 1 here: https://arxiv.org/pdf/1704.00028.pdf

In order to do so, we basically need an interface like the tf.gradient one:
https://www.tensorflow.org/versions/r0.10/api_docs/python/train/gradient_computation

It would be nice to have support for this. The Jacobian builder branch is a first stab at this, but it's far from finished:
https://github.com/clab/dynet/tree/jacobian-builder
